File: Crio/crio_programming_interview_problems-master/ConvertSortedArrayToBinarySearchTree/ConvertSortedArrayToBinarySearchTree.cpp
#include <bits/stdc++.h>
#include "FastIO.hpp"
#include "ReadMatrix.hpp"
#include "PrintMatrix.hpp"
#include "TreeNode.hpp"

using namespace std;

class ConvertSortedArrayToBinarySearchTree {
    public:
        // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
        TreeNode* tree(vector<int>& nums, int l, int r) {
            if (l > r) return nullptr;
            int mid = (l + r) / 2;
            TreeNode* root = new TreeNode(nums[mid]);
            root->left = tree(nums, l, mid - 1);
            root->right = tree(nums, mid + 1, r);
            return root;
        }
        // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

        TreeNode* sortedArrayToBST(vector<int>& nums) {
            // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
            return tree(nums,0,nums.size()-1);
            // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
        }
};

int main() {
    FastIO();
    int n;
    cin >> n;
    vector<int> nums;
    ReadMatrix<int>().OneDMatrix(n, nums);
    TreeNode* root = ConvertSortedArrayToBinarySearchTree().sortedArrayToBST(nums);
    string result = TreeNode().treeNodeToString(root);
    cout << result;
    return 0;
}

File: Crio/crio_programming_interview_problems-master/ConvertSortedArrayToBinarySearchTree/ConvertSortedArrayToBinarySearchTree.java
import crio.ds.Tree.TreeNode;
import java.io.*;
import java.util.*;

public class ConvertSortedArrayToBinarySearchTree {
    // Implement your solution by completing the below function
    public TreeNode sortedArrayToBST(int[] num) {
        // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
        if (num.length == 0) {
            return null;
        }
        TreeNode head = helper(num, 0, num.length - 1);
        return head;
        // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

        // CRIO_UNCOMMENT_START_MODULE_L1_PROBLEMS
        //return new TreeNode(-1);
        // CRIO_UNCOMMENT_END_MODULE_L1_PROBLEMS
    }

    // CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    public TreeNode helper(int[] num, int low, int high) {
        if (low > high) { // Done
            return null;
        }
        int mid = (low + high) / 2;
        TreeNode node = new TreeNode(num[mid]);
        node.left = helper(num, low, mid - 1);
        node.right = helper(num, mid + 1, high);
        return node;
    }
    // CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

    public static void main(String[] args) throws IOException {
        Scanner scanner = new Scanner(System.in);

        int[] array = new int[scanner.nextInt()];
        for (int i = 0; i < array.length; i++)
            array[i] = scanner.nextInt();
        scanner.close();

        TreeNode treeRoot = new ConvertSortedArrayToBinarySearchTree().sortedArrayToBST(array);
        String treeAsArray = TreeNode.treeNodeToString(treeRoot);
        System.out.print(treeAsArray);
    }
}